doc/dev/static_type_checking.md

```diff
@@ -0,0 +1,894 @@
+promises to return the same `T` type passed in.
+
+You can restrict a `TypeVar` to several specific types by adding the types as positional arguments:
+
+```python
+from typing import Sequence, TypeVar
+
+T = TypeVar("T", int, str)
+
+
+def pick(p: Sequence[T]) -> T:
+    return random_pick(p)
+```
+
+Here the type checker will only expect types of `int` or `str` for `T`.
+
+There are also arguments available to set the upper boundary or variance in a `TypeVar` -- please see
+the [reference docs](https://docs.python.org/3/library/typing.html#typing.TypeVar) for more information.
+
+### Use typing.Protocol for structural subtyping
+
+[PEP 544](https://peps.python.org/pep-0544/) introduced the `Protocol` type. The `Protocol` type helps Python support _
+structural subtyping_ or "static duck typing". In other words, when looking at whether two types are compatible, the
+actual structure and capabilities are more important than the _name_ of type.
+
+A `Protocol` type can be defined by one or more methods, and a type checker will check that those methods are
+implemented where a protocol type is specified.
+
+Below we create a `SupportsFly` protocol class which expects a `fly()` method on all compatible types being passed
+into `ascend()`.
+
+```python
+# from typing import Protocol  # Python >=3.8
+from typing_extensions import Protocol
+
+
+class SupportsFly(Protocol):
+    def fly(self) -> None: ...
+
+
+class Butterfly:
+    def fly(self) -> None: ...
+
+    def sunbathe(self) -> None: ...
+
+
+class Plane:
+    def fly(self) -> None: ...
+
+
+class Penguin:
+    def waddle(self) -> None: ...
+
+
+def ascend(f: SupportsFly) -> None:
+    f.fly()
+
+
+p = Plane()
+ascend(p)  # OK
+
+b = Butterfly()
+ascend(b)  # OK
+
+p = Penguin()
+ascend(p)  # Mypy complains below
+```
+
+Running mypy shows expected output:
+
+```cmd
+main.py:23: error: Argument 1 to "ascend" has incompatible type "Penguin"; expected "SupportsFly"
+Found 1 error in 1 file (checked 1 source file)
+```
+
+Because Penguin does not implement the `SupportsFly` Protocol is called out as an incompatible type. However, any type
+that is consistent with, or implements all the methods defined by the `SupportsFly` protocol, is acceptable to pass
+into `ascend()`.
+
+A key difference between Protocols and abstract classes (abc) is that protocols work at typing time while abcs work at
+runtime. A class does not need to _inherit_ from Protocol like an abc, the type checker will simply enforce its usage at
+typing time.
+
+A common Python SDK Protocol type we use
+is [azure.core.credentials.TokenCredential](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/core/azure-core/azure/core/credentials.py)
+. TokenCredential requires that a type implements the `get_token` method for it to be consistent with the Protocol.
+
+See more on Protocols in the [reference documentation](https://docs.python.org/3/library/typing.html#typing.Protocol).
```

Comment:
Was runtime_checkable backported? If yes we should mention it:

https://docs.python.org/3/library/typing.html#typing.runtime_checkable